File: backend/backend/models/metrics.py
from typing import Dict, List, Optional
from pydantic import BaseModel


class CollectorPerf(BaseModel):
    name: str
    # type: str
    call_cnt: int
    err_cnt: int
    max_us: float
    mean_us: float
    min_us: float
    variance: float
    # metrics_collected: int
    # avg_collection_time: float
    # errors: int
    # last_error: Optional[str] = None


class WriterPerf(BaseModel):
    name: str
    call_cnt: int
    err_cnt: int
    max_us: float
    mean_us: float
    min_us: float
    variance: float
    # metrics_written: int
    # avg_write_time: float
    # buffer_size: int
    # errors: int


class WriterInfo(BaseModel):
    name: str
    type: str
    config: Dict
    status: str
    metrics_written: int


class ServiceMetrics(BaseModel):
    service_id: str
    service_name: str
    collectors: List[CollectorPerf]
    writers: List[WriterPerf]


class PrometheusMetrics(BaseModel):
    content: str

File: frontend/src/api/metrics.ts
import { apiClient } from './config'

// export interface CollectorPerf {
//   name: string
//   type: string
//   metrics_collected: number
//   avg_collection_time: number
//   errors: number
//   last_error?: string
// }

export interface CollectorPerf {
  name: string
  // type: string
  call_cnt: number
  err_cnt: number
  max_us: number
  mean_us: number
  min_us: number
  variance: number
}

// export interface WriterPerf {
//   name: string
//   metrics_written: number
//   avg_write_time: number
//   buffer_size: number
//   errors: number
// }

export interface WriterPerf {
  name: string
  call_cnt: number
  err_cnt: number
  max_us: number
  mean_us: number
  min_us: number
  variance: number
}

export interface WriterInfo {
  name: string
  type: string
  config: Record<string, unknown>
  status: string
  metrics_written: number
}

export interface ServiceMetrics {
  service_id: string
  service_name: string
  collectors: CollectorPerf[]
  writers: WriterPerf[]
}

export interface PrometheusMetrics {
  content: string
}

export interface RegistryMetrics {
  registry_health: {
    status: string
    version: string
    uptime?: string
  }
  registry_stats: {
    total_services: number
    healthy_services: number
    unhealthy_services: number
    active_services: number
  }
}

export const metricsApi = {
  async getCollectorPerformance(serviceId: string): Promise<CollectorPerf[]> {
    const response = await apiClient.get(`/metrics/services/${serviceId}/collectors`)
    return response.data
  },

  async getWriterPerformance(serviceId: string): Promise<WriterPerf[]> {
    const response = await apiClient.get(`/metrics/services/${serviceId}/writers`)
    return response.data
  },

  async getWriterInfo(serviceId: string, writerName: string): Promise<WriterInfo> {
    const response = await apiClient.get(`/metrics/services/${serviceId}/writers/${writerName}`)
    return response.data
  },

  async getAllMetrics(serviceId: string): Promise<ServiceMetrics> {
    const response = await apiClient.get(`/metrics/services/${serviceId}/all`)
    return response.data
  },

  async getPrometheusMetrics(serviceId: string): Promise<string> {
    const response = await apiClient.get(`/metrics/services/${serviceId}/prometheus`)
    return response.data.content
  },

  async getRegistryMetrics(): Promise<RegistryMetrics> {
    const response = await apiClient.get('/metrics/registry')
    return response.data
  },
}
